src/net/socket/tcp_sock.c

```diff
@@ -374,39 +374,75 @@ static int tcp_accept(struct sock *sk, struct sockaddr *addr,
 	return 0;
 }
 
-static int tcp_write(struct tcp_sock *tcp_sk, void *buff, size_t len) {
+static int tcp_write(struct tcp_sock *tcp_sk, struct msghdr * msg) {
+	size_t iov_len;
+	int full_len;
 	void *pb;
 	struct sk_buff *skb;
 	int ret;
+	size_t sbk_len;
+	int tran_len;
+	int cp_len;
+	int cp_off;
+	int i;
+
+	full_len = 0;
+	tran_len = 0;
+	for (i = 0; i < msg->msg_iovlen; i ++) {
+		full_len += msg->msg_iov[i].iov_len;
+	}
 
-	pb = buff;
-	while (len != 0) {
-		/* Previous comment: try to send wholly msg
-		 * We must pass no more than 64k bytes to underlaying IP level */
-		size_t bytes = min(len, IP_MAX_PACKET_LEN - MAX_HEADER_SIZE);
-		skb = NULL; /* alloc new pkg */
+	iov_len = 0;
+	sbk_len = 0;
 
-		ret = alloc_prep_skb(tcp_sk, 0, &bytes, &skb);
-		if (ret != 0) {
-			break;
+	for (i = 0; i < msg->msg_iovlen; ) {
+		if (0 == iov_len) {
+			pb = msg->msg_iov[i].iov_base;
+			iov_len = msg->msg_iov[i].iov_len;
+			i++;
 		}
 
-		log_debug("sending len %d", bytes);
+		if (0 == sbk_len) {
+			cp_off = 0;
+			sbk_len = min((full_len - tran_len), IP_MAX_PACKET_LEN - MAX_HEADER_SIZE);
+			skb = NULL; /* alloc new pkg */
+
+			ret = alloc_prep_skb(tcp_sk, 0, &sbk_len, &skb);
+			if (ret != 0) {
+				break;
+			}
 
-		tcp_build(skb->h.th,
+			tcp_build(skb->h.th,
 				sock_inet_get_dst_port(to_sock(tcp_sk)),
 				sock_inet_get_src_port(to_sock(tcp_sk)),
 				TCP_MIN_HEADER_SIZE, tcp_sk->self.wind.value);
+		}
+
+		cp_len = min(iov_len, sbk_len);
+
+		memcpy(((void *)(skb->h.th + 1) + cp_off), pb, cp_len);
+		iov_len -= cp_len;
+		pb += cp_len;
+		sbk_len -= cp_len;
+		tran_len += cp_len;
+		cp_off += cp_len;
+
+		if (0 == sbk_len) {
+			/* Fill TCP header */
+			skb->h.th->psh = (iov_len == 0);
+			tcp_set_ack_field(skb->h.th, tcp_sk->rem.seq);
+			send_seq_from_sock(tcp_sk, skb);
+			cp_off = 0;
+		}
+	}
 
-		memcpy(skb->h.th + 1, pb, bytes);
-		pb += bytes;
-		len -= bytes;
+	if (0 != cp_off) {
```

Comment:
Is it normal behavior that cp_off is not zero? I added `assert(0)` inside this if branch and didn't get it under telnet and httpd.